metric.go

```diff
@@ -210,26 +233,72 @@ func (m *constMetric) Write(out *dto.Metric) errors.WithContext {
 	return nil
 }
 
-func makeLabelPairs(desc MetricDesc, labelValues []string) []*dto.LabelPair {
+func parseJsonLabels(desc MetricDesc, labels string) []*dto.LabelPair {
+	var userLabels []*dto.LabelPair
+	var jsonLabels map[string]string
+
+	config := desc.GlobalConfig()
+	maxJsonLabels := 0
+	if (config != nil) {
+		maxJsonLabels = config.MaxJsonLabels
+	}
+
+	err := json.Unmarshal([]byte(labels), &jsonLabels)
+	// errors are logged but ignored
+	if err != nil {
+		log.Warningf("[%s] Failed to parse JSON labels returned by query - %s", desc.LogContext(), err)
+	} else {
+		userLabelsMax := int(math.Min(float64(len(jsonLabels)), float64(maxJsonLabels)))
+		userLabels = make([]*dto.LabelPair, userLabelsMax)
+
+		idx := 0
+		for name, value := range jsonLabels {
+			// limit label count
+			if idx >= maxJsonLabels {
```

Comment:
Have you actually tested this with a limit lower than the number of labels? I'm pretty sure it won't work as expected, I just don't know how exactly it will fail. (Most likely the first metric will be exported and all others, with different sets of labels will not. But even if that is the case, I'm not even sure whether the set of labels would be determined on every scrape or only on the first one.)

Reply:
yes, i've tested it. it truncates randomly so most metrics go trough, but some not and prometheus client logs a warning about inconsistent labels. I've added one more warning that truncation occurs, to make it better traceable.
I could maybe sort the labels, but I find it too big performance hit and I think that the limit will be almost never reached, its there just for sure